# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


task 1

H₀: μ = 123,7 (середній час покупки на сайті після редизайну не змінився)

H₁: μ > 123,7 (середній час покупки на сайті після редизайну збільшився)

Тип тесту: правосторонній (one-tailed)

In [34]:
import numpy as np
from scipy import stats
time_of_purchase = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7
alpha = 0.10
x_bar = np.mean(time_of_purchase)
s = np.std(time_of_purchase, ddof=1)
n = len(time_of_purchase)

# Крок 2. Обчислимо t-статистику вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Крок 3. Критичне значення для лівостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1-alpha, df)

# Крок 4. p-value
p_value = 1-stats.t.cdf(t_stat, df)

# Крок 5. Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості 0.1


In [35]:
res = stats.ttest_1samp(time_of_purchase, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}")
if res.pvalue < alpha:
    print(f"Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

scipy: t = 2.069, p-value = 0.04202
Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості 0.1


In [36]:
t_crit = stats.t.ppf(1-alpha, res.df)

if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: середній час покупки на сайті після редизайну збільшився на рівні значущості 0.1


Висновок: середній час покупки на сайті після редизайну збільшився на рівні значущості 0.1

При зміні рівня значущості на 0.05 рішення про відхилення H0 не змінилось

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [40]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311   # кількість "успіхів"
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)

# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print("🔹 Тест для однієї пропорції")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

🔹 Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


Оскільки p-value = 0.0106 < 0.03, ми відхиляємо нульову гіпотезу H₀.
Це означає, що є статистично значущі докази того, що частка дорослих, які впізнають логотип компанії після рекламної кампанії, перевищує 23%.

Довірчий інтервал(0.231, 0.286) означає, що з імовірністю 97% справжня частка лежить у межах від 23.1% до 28.6%.

Справжня частка людей, які впізнають логотип, знаходиться приблизно між 23.1% і 28.6%

23% є "поза нормальним діапазоном" для реальних значень.
Отже, воно малоймовірне як істинне значення (тому відхиляємо H₀)